examples/chainrules.jl

```diff
@@ -0,0 +1,156 @@
+"""
+function unit_commitment(load1_demand, load2_demand, gen_costs, noload_costs; model = Model(() -> diff_optimizer(Clp.Optimizer)))
+    ## Problem data
+    unit_codes = [1, 2] # Generator identifiers
+    load_names = ["Load1", "Load2"] # Load identifiers
+    n_periods = 4 # Number of time periods
+    Pmin = Dict(1 => fill(0.5, n_periods), 2 => fill(0.5, n_periods)) # Minimum power output (pu)
+    Pmax = Dict(1 => fill(3.0, n_periods), 2 => fill(3.0, n_periods)) # Maximum power output (pu)
+    RR = Dict(1 => 0.25, 2 => 0.25) # Ramp rates (pu/min)
+    P0 = Dict(1 => 0.0, 2 => 0.0) # Initial power output (pu)
+    D = Dict("Load1" => load1_demand, "Load2" => load2_demand) # Demand (pu)
+    Cp = Dict(1 => gen_costs[1], 2 => gen_costs[2]) # Generation cost coefficient ($/pu)
+    Cnl = Dict(1 => noload_costs[1], 2 => noload_costs[2]) # No-load cost ($)
+
+    ## Variables
+    # Note: u represents the activation of generation units.
+    # Would be binary in the typical UC problem, relaxed here to u ∈ [0,1]
+    # for a linear relaxation.
+    @variable(model, 0 <= u[g in unit_codes, t in 1:n_periods] <= 1) # Commitment
+    @variable(model, p[g in unit_codes, t in 1:n_periods] >= 0) # Power output (pu)
+    
+    ## Constraints
+    
+    # Energy balance
+    @constraint(
+        model,
+        energy_balance_cons[t in 1:n_periods],
+        sum(p[g, t] for g in unit_codes) == sum(D[l][t] for l in load_names),
+    )
+    
+    # Generation limits
+    @constraint(model, [g in unit_codes, t in 1:n_periods], Pmin[g][t] * u[g, t] <= p[g, t])
+    @constraint(model, [g in unit_codes, t in 1:n_periods], p[g, t] <= Pmax[g][t] * u[g, t])
+    
+    # Ramp rates
+    @constraint(model, [g in unit_codes, t in 2:n_periods], p[g, t] - p[g, t - 1] <= 60 * RR[g])
+    @constraint(model, [g in unit_codes], p[g, 1] - P0[g] <= 60 * RR[g])
+    @constraint(model, [g in unit_codes, t in 2:n_periods], p[g, t - 1] - p[g, t] <= 60 * RR[g])
+    @constraint(model, [g in unit_codes], P0[g] - p[g, 1] <= 60 * RR[g])
+    
+    # Objective
+    @objective(
+        model,
+        Min,
+        sum((Cp[g] * p[g, t]) + (Cnl[g] * u[g, t]) for g in unit_codes, t in 1:n_periods),
+    )
+    
+    optimize!(model)
+    @assert termination_status(model) == MOI.OPTIMAL
+    # converting to dense matrix
+    return JuMP.value.(p.data)
+end
+
+@show unit_commitment([1.0, 1.2, 1.4, 1.6], [1.0, 1.2, 1.4, 1.6], [1000.0, 1500.0], [500.0, 1000.0])
+
+# Forward differentiation rule for the solution map of the unit commitment problem
+# taking in input perturbations on the input parameters and returning perturbations propagated to the result
+function ChainRulesCore.frule((_, Δload1_demand, Δload2_demand, Δgen_costs, Δnoload_costs), ::typeof(unit_commitment), load1_demand, load2_demand, gen_costs, noload_costs)
+    # creating the UC model with a DiffOpt optimizer wrapper around Clp
+    model = Model(() -> diff_optimizer(Clp.Optimizer))
```

Comment:
We need to also pass the kwargs in: so probably can do:
```suggestion
function ChainRulesCore.frule((_, Δload1_demand, Δload2_demand, Δgen_costs, Δnoload_costs), ::typeof(unit_commitment), load1_demand, load2_demand, gen_costs, noload_costs; model=Model(Clp.Optimizer))
    # creating the UC model with a DiffOpt optimizer wrapper around Clp
    model = Model(() -> diff_optimizer(optimizer(model)))
```

Reply:
one issue here is that `backend(.)` only gives you the top-most backend layer, so you add two layers of CachingOptimizer